### ISSUE TITLE
Possible computeRayKeys method improvement

### ISSUE THREAD

**Issue by OndrejMasopust** [OctoMap/octomap#324]
While looking at how the computeRayKeys is done, I came across this line.
https://github.com/OctoMap/octomap/blob/7a9ab4a357860f7263769cce9dd45107ec8b2c4a/octomap/include/octomap/OcTreeBaseImpl.hxx#L549
Would it be possible to actually not reset the rays vector? It would speed up the method and allow the user to add new keys into an already existing KeyRay object without reseting it. Is there something I don't know about which makes this change unacceptable?

**Comment by ahornung** [comment]
The mentioned code does not reset a vector, it resets the KeyRay storage. The KeyRay class holds an internal vector, which is not reset through this call but only resets the iterator to the start.

**Comment by OndrejMasopust** [comment]
Yes, maybe I just didn't describe my thought very well. My point was that if a programmer wants to add new keys to a KeyRay instance, which is not empty, this method will overwrite the already stored keys, right?

**Comment by ahornung** [comment]
Yes, the linked method will always start writing at the beginning. The intended usage is to compute the keys along the ray starting at the origin, and for this it needs to start at the beginning (and thus reset the ray storage).

**Comment by OndrejMasopust** [comment]
Ok, my idea was concerning situations when the programmer wants to process nodes from multiple rays. It is convenient to store these node keys in one KeyRay. I thought about it in such a way that these nodes will be first obtained by multiple calls of the computeRayKeys method. But since this method resets the KeyRay, it is not possible to store node keys from multiple rays in one KeyRay. Possible solution to this is to use more KeyRay instances and then somehow "fuse" them with the addKey method. But this would be easier if the computeRayKeys method didn't reset the KeyRay. If the programmer needed to overwrite the KeyRay before calling computeRayKeys, it would be then his responsibility to do so before making the call. So this is my idea for you to consider if it is useful.

I faced a situation like this when I created this issue. But now I found out a better way how to solve my problem regardless of whether the computeRayKeys method resets the ray storage or not. So rather than a "need", this is now more of an idea or a suggestion for possible future upgrade.
